File: swift/src/SHA1.swift
/// Pure Swift SHA-1 implementation
/// Parity: libgit2 uses SHA-1 for object IDs (see src/util/hash/sha1/)
import Foundation

public struct SHA1 {
    private var h0: UInt32 = 0x67452301
    private var h1: UInt32 = 0xEFCDAB89
    private var h2: UInt32 = 0x98BADCFE
    private var h3: UInt32 = 0x10325476
    private var h4: UInt32 = 0xC3D2E1F0

    private var buffer = [UInt8]()
    private var totalLength: UInt64 = 0

    public init() {}

    public mutating func update(_ data: [UInt8]) {
        buffer.append(contentsOf: data)
        totalLength += UInt64(data.count)

        while buffer.count >= 64 {
            let block = Array(buffer.prefix(64))
            processBlock(block)
            buffer.removeFirst(64)
        }
    }

    public mutating func update(_ data: Data) {
        update(Array(data))
    }

    public mutating func update(_ string: String) {
        update(Array(string.utf8))
    }

    public mutating func finalize() -> [UInt8] {
        // Pad message
        var padded = buffer
        padded.append(0x80)

        while padded.count % 64 != 56 {
            padded.append(0x00)
        }

        // Append original length in bits as big-endian 64-bit
        let bitLength = totalLength * 8
        for i in (0..<8).reversed() {
            padded.append(UInt8((bitLength >> (i * 8)) & 0xFF))
        }

        // Process remaining blocks
        var offset = 0
        while offset < padded.count {
            let block = Array(padded[offset..<offset + 64])
            processBlock(block)
            offset += 64
        }

        // Produce digest
        var digest = [UInt8](repeating: 0, count: 20)
        for (i, h) in [h0, h1, h2, h3, h4].enumerated() {
            digest[i * 4]     = UInt8((h >> 24) & 0xFF)
            digest[i * 4 + 1] = UInt8((h >> 16) & 0xFF)
            digest[i * 4 + 2] = UInt8((h >> 8) & 0xFF)
            digest[i * 4 + 3] = UInt8(h & 0xFF)
        }
        return digest
    }

    private mutating func processBlock(_ block: [UInt8]) {
        var w = [UInt32](repeating: 0, count: 80)

        // Load 16 words from block (big-endian)
        for i in 0..<16 {
            w[i] = UInt32(block[i * 4]) << 24
                | UInt32(block[i * 4 + 1]) << 16
                | UInt32(block[i * 4 + 2]) << 8
                | UInt32(block[i * 4 + 3])
        }

        // Extend to 80 words
        for i in 16..<80 {
            w[i] = rotateLeft(w[i-3] ^ w[i-8] ^ w[i-14] ^ w[i-16], by: 1)
        }

        var a = h0, b = h1, c = h2, d = h3, e = h4

        for i in 0..<80 {
            let f: UInt32
            let k: UInt32

            switch i {
            case 0..<20:
                f = (b & c) | ((~b) & d)
                k = 0x5A827999
            case 20..<40:
                f = b ^ c ^ d
                k = 0x6ED9EBA1
            case 40..<60:
                f = (b & c) | (b & d) | (c & d)
                k = 0x8F1BBCDC
            default:
                f = b ^ c ^ d
                k = 0xCA62C1D6
            }

            let temp = rotateLeft(a, by: 5) &+ f &+ e &+ k &+ w[i]
            e = d
            d = c
            c = rotateLeft(b, by: 30)
            b = a
            a = temp
        }

        h0 = h0 &+ a
        h1 = h1 &+ b
        h2 = h2 &+ c
        h3 = h3 &+ d
        h4 = h4 &+ e
    }

    private func rotateLeft(_ value: UInt32, by count: UInt32) -> UInt32 {
        (value << count) | (value >> (32 - count))
    }

    /// Convenience: hash data in one call
    public static func hash(_ data: [UInt8]) -> [UInt8] {
        var sha = SHA1()
        sha.update(data)
        return sha.finalize()
    }

    /// Convenience: hash string
    public static func hash(_ string: String) -> [UInt8] {
        hash(Array(string.utf8))
    }
}

// MARK: - OID SHA-1 Extensions

extension OID {
    /// Create an OID by hashing data with SHA-1 (git object style)
    public static func hash(type: ObjectType, data: [UInt8]) -> OID {
        let typeName: String
        switch type {
        case .commit: typeName = "commit"
        case .tree:   typeName = "tree"
        case .blob:   typeName = "blob"
        case .tag:    typeName = "tag"
        }

        let header = "\(typeName) \(data.count)\0"
        var sha = SHA1()
        sha.update(Array(header.utf8))
        sha.update(data)
        return OID(raw: sha.finalize())
    }

    /// SHA-1 digest length in bytes
    public static let sha1Length = 20

    /// SHA-1 hex string length
    public static let sha1HexLength = 40

    /// Whether this OID is all zeros
    public var isZero: Bool {
        raw.allSatisfy { $0 == 0 }
    }

    /// Zero OID
    public static let zero = OID(raw: [UInt8](repeating: 0, count: sha1Length))
}

File: rust/src/sha1.rs
//! SHA-1 implementation using the `sha1` crate for hardware-accelerated hashing.
//! Parity: libgit2 uses SHA-1 for object IDs (see src/util/hash/sha1/)

use sha1_crate::{Sha1, Digest};

pub struct SHA1 {
    hasher: Sha1,
}

impl SHA1 {
    pub fn new() -> Self {
        Self { hasher: Sha1::new() }
    }

    pub fn update(&mut self, data: &[u8]) {
        self.hasher.update(data);
    }

    pub fn finalize(self) -> [u8; 20] {
        self.hasher.finalize().into()
    }

    /// Hash data in one call
    pub fn hash(data: &[u8]) -> [u8; 20] {
        let mut sha = SHA1::new();
        sha.update(data);
        sha.finalize()
    }
}

impl Default for SHA1 {
    fn default() -> Self {
        Self::new()
    }
}

/// Build git object header ("type size\0") into a stack buffer, returns valid slice length
pub fn build_object_header(obj_type: crate::ObjectType, data_len: usize, buf: &mut [u8; 32]) -> usize {
    let type_bytes: &[u8] = match obj_type {
        crate::ObjectType::Commit => b"commit ",
        crate::ObjectType::Tree => b"tree ",
        crate::ObjectType::Blob => b"blob ",
        crate::ObjectType::Tag => b"tag ",
    };
    let mut pos = type_bytes.len();
    buf[..pos].copy_from_slice(type_bytes);
    // Write decimal size
    if data_len == 0 {
        buf[pos] = b'0';
        pos += 1;
    } else {
        let start = pos;
        let mut v = data_len;
        while v > 0 {
            buf[pos] = b'0' + (v % 10) as u8;
            v /= 10;
            pos += 1;
        }
        buf[start..pos].reverse();
    }
    buf[pos] = 0; // null terminator
    pos += 1;
    pos
}

impl crate::OID {
    /// Create an OID by hashing data with SHA-1 (git object style)
    pub fn hash_object(obj_type: crate::ObjectType, data: &[u8]) -> Self {
        let mut header_buf = [0u8; 32];
        let header_len = build_object_header(obj_type, data.len(), &mut header_buf);
        let mut sha = SHA1::new();
        sha.update(&header_buf[..header_len]);
        sha.update(data);
        Self::from_bytes(sha.finalize().to_vec())
    }

    /// SHA-1 digest length in bytes
    pub const SHA1_LENGTH: usize = 20;

    /// SHA-1 hex string length
    pub const SHA1_HEX_LENGTH: usize = 40;

    /// Whether this OID is all zeros
    pub fn is_zero(&self) -> bool {
        self.raw().iter().all(|&b| b == 0)
    }

    /// Zero OID
    pub fn zero() -> Self {
        Self::from_bytes(vec![0u8; Self::SHA1_LENGTH])
    }

    /// Create OID from raw bytes (direct, no hex round-trip)
    pub fn from_bytes(raw: Vec<u8>) -> Self {
        Self::new(raw)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_sha1_empty() {
        let digest = SHA1::hash(b"");
        let hex: String = digest.iter().map(|b| format!("{:02x}", b)).collect();
        assert_eq!(hex, "da39a3ee5e6b4b0d3255bfef95601890afd80709");
    }

    #[test]
    fn test_sha1_hello() {
        let digest = SHA1::hash(b"hello");
        let hex: String = digest.iter().map(|b| format!("{:02x}", b)).collect();
        assert_eq!(hex, "aaf4c61ddcc5e8a2dabede0f3b482cd9aea9434d");
    }

    #[test]
    fn test_sha1_git_blob() {
        // git hash-object equivalent: "hello\n" as blob
        let data = b"hello\n";
        let oid = crate::OID::hash_object(crate::ObjectType::Blob, data);
        assert_eq!(oid.hex(), "ce013625030ba8dba906f756967f9e9ca394464a");
    }

    #[test]
    fn test_oid_zero() {
        let z = crate::OID::zero();
        assert!(z.is_zero());
        assert_eq!(z.hex(), "0000000000000000000000000000000000000000");
    }
}
